internal/integration/entity_synthesis.go

```diff
@@ -0,0 +1,169 @@
+// Package integration ...
+// Copyright 2021 New Relic Corporation. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package integration
+
+import (
+	"strings"
+
+	sdk_metadata "github.com/newrelic/infra-integrations-sdk/v4/data/metadata"
+	"github.com/newrelic/nri-prometheus/internal/pkg/labels"
+)
+
+// Synthesizer group of rules to synthesis entities
+type Synthesizer struct {
+	rulesByConditions map[Condition]EntityRule
+}
+
+// NewSynthesizer initialize and return a Synthesizer from a set of EntityRules
+func NewSynthesizer(entitySynthesisDefinitions []SynthesisDefinition) Synthesizer {
+	s := Synthesizer{
+		rulesByConditions: make(map[Condition]EntityRule),
+	}
+	for _, ed := range entitySynthesisDefinitions {
+		ed.tagRules = ed.Tags.getTagRules()
+		s.addConditions(ed.EntityRule)
+
+		for _, er := range ed.Rules {
+			er.EntityType = ed.EntityType
+			er.tagRules = append(er.Tags.getTagRules(), ed.EntityRule.tagRules...)
+			s.addConditions(er)
+		}
+	}
+	return s
+}
+func (s *Synthesizer) addConditions(rule EntityRule) {
+	if !rule.isValid() {
+		return
+	}
+	for _, c := range rule.Conditions {
+		s.rulesByConditions[c] = rule
+	}
+}
+
+// SynthesisDefinition contains rules to synthesis entities from metrics
+type SynthesisDefinition struct {
+	EntityRule `mapstructure:",squash"`
+	Rules      []EntityRule `mapstructure:"rules"`
+}
+
+// EntityRule contains rules to synthesis an entity
+type EntityRule struct {
+	EntityType string      `mapstructure:"type"`
+	Identifier string      `mapstructure:"identifier"`
+	Name       string      `mapstructure:"name"`
+	Conditions []Condition `mapstructure:"conditions"`
+	Tags       Tags        `mapstructure:"tags"`
+	tagRules   []tagRule
+}
+
+func (rule EntityRule) isValid() bool {
+	return rule.Identifier != "" && rule.Name != "" && rule.EntityType != ""
+}
+
+// Condition contains parameters used to match entities from metrics
+type Condition struct {
+	Attribute string `mapstructure:"attribute"`
+	Prefix    string `mapstructure:"prefix"`
+	Value     string `mapstructure:"value"`
+}
+
+// match evaluates the condition for a particular existing attribute value.
+func (c Condition) match(attribute string) bool {
+	if c.Value != "" {
+		return c.Value == attribute
+	}
+	// this returns true if c.Prefix is "" and is ok since the attribute exists
+	return strings.HasPrefix(attribute, c.Prefix)
+}
+
+// Tags key value attributes
+type Tags map[string]map[string]interface{}
+
+type tagRule struct {
+	name          string
+	entityTagName string
+}
```

Comment:
If I understood this correctly, `name` is the name of the attribute that contains the to-be-created tag value, and `entityTagName` is the name of this same tag after applying the renaming logic in the entity synthesis rule. Is this correct? If so, it might be better to clarify this either with a comment, or in the names of these fields:
```suggestion
type tagRule struct {
	sourceAttribute string
	newTagName      string
}
```

Reply:
I followed the entity synthesis protocol naming here, I think it will make easier to read if you are already familiar with the protocol , which is indeed needed for understand this package. I fear that doing more renaming will cause more confusion at the end.